lib/src/source/hosted.dart

```diff
@@ -557,22 +564,28 @@ class BoundHostedSource extends CachedSource {
       throw PackageNotFoundException(
           'Package $packageName has no version $version');
     }
+    final parsedDescription = source._parseDescription(id.description);
+    final server = parsedDescription.last;
+
     var url = versionInfo.archiveUrl;
-    if (url == null) {
-      // To support old servers that has no archive_url we fall back to the
-      // hard-coded path.
-      final parsedDescription = source._parseDescription(id.description);
-      final server = parsedDescription.last;
-      url = Uri.parse('$server/packages/$packageName/versions/$version.tar.gz');
-    }
+    // To support old servers that has no archive_url we fall back to the
+    // hard-coded path.
+    url ??= Uri.parse('$server/packages/$packageName/versions/$version.tar.gz');
     log.io('Get package from $url.');
     log.message('Downloading ${log.bold(id.name)} ${id.version}...');
 
     // Download and extract the archive to a temp directory.
     await withTempDir((tempDirForArchive) async {
       var archivePath =
           p.join(tempDirForArchive, '$packageName-$version.tar.gz');
-      var response = await httpClient.send(http.Request('GET', url));
+      var response = await withAuthenticatedClient(
+        systemCache,
+        server,
+        (client) {
+          return client.send(http.Request('GET', url));
+        },
+        alsoMatches: [archivePath],
```

Comment:
What is this?

Reply:
I'll probably replace `alsoMatches` with just `serverUrls` argument that accepts list of strings. It will be used to choose a credential that matches all the server urls. It's there to ensure credentials are not leaked to 3rd parties when archives are not hosted on the same authenticated server.

..And I just noticed that, `archivePath` has to be `url` instead, sorry for that.